src/nnet3/nnet-simple-component.h

```diff
@@ -714,6 +714,19 @@ class NaturalGradientAffineComponent: public AffineComponent {
   // in Info().
   double max_change_scale_stats_;
 
+  /// The following function is only called if max_change_per_sample_ > 0, it returns a
+  /// scaling factor alpha <= 1.0 (1.0 in the normal case) that enforces the
+  /// "max-change" constraint.  "in_products" is the inner product with itself
+  /// of each row of the matrix of preconditioned input features; "out_products"
+  /// is the same for the output derivatives.  gamma_prod is a product of two
+  /// scalars that are output by the preconditioning code (for the input and
+  /// output), which we will need to multiply into the learning rate.
+  /// out_products is a pointer because we modify it in-place.
```

Comment:
Is it possible to provide a description of what max-change intends to do in detail ?
Something like a reference to a paper or at least some equations here describing whats being done and why.